File: figma_mcp_client.py
#!/usr/bin/env python3
import requests
import json
import re

MCP_URL = "http://127.0.0.1:3845/mcp"

def parse_sse_response(text):
    """Parse Server-Sent Events format"""
    # Extract JSON from SSE format
    if "data: " in text:
        for line in text.split('\n'):
            if line.startswith('data: '):
                return json.loads(line[6:])
    # Try parsing as regular JSON
    try:
        return json.loads(text)
    except:
        return None

def make_session_request(method, params=None, request_id=1):
    """Make a single request with a new session"""
    headers = {
        "Content-Type": "application/json",
        "Accept": "application/json, text/event-stream"
    }

    # First, initialize
    init_payload = {
        "jsonrpc": "2.0",
        "id": 0,
        "method": "initialize",
        "params": {
            "protocolVersion": "2024-11-05",
            "capabilities": {},
            "clientInfo": {
                "name": "claude-code",
                "version": "1.0.0"
            }
        }
    }

    session = requests.Session()
    init_response = session.post(MCP_URL, json=init_payload, headers=headers)
    print(f"Init response status: {init_response.status_code}")

    # Now make the actual request
    payload = {
        "jsonrpc": "2.0",
        "id": request_id,
        "method": method
    }
    if params:
        payload["params"] = params

    response = session.post(MCP_URL, json=payload, headers=headers)
    return response.text, session

# Get tools list
print("=== Fetching Available Tools ===")
tools_response, tools_session = make_session_request("tools/list", request_id=1)
print(tools_response)
print()

tools_json = parse_sse_response(tools_response)
if tools_json and "result" in tools_json:
    tools = tools_json.get("result", {}).get("tools", [])
    print(f"\nFound {len(tools)} tools:")
    for tool in tools:
        print(f"\n  Name: {tool['name']}")
        print(f"  Description: {tool.get('description', 'No description')}")
        if 'inputSchema' in tool:
            print(f"  Input schema: {json.dumps(tool['inputSchema'], indent=4)}")

    # Try to call relevant tools
    print("\n=== Attempting to Get Design Specifications ===")
    for tool in tools:
        tool_name = tool['name']
        print(f"\n--- Calling: {tool_name} ---")

        # Prepare arguments based on the tool
        args = {}

        tool_call_response, _ = make_session_request("tools/call", {
            "name": tool_name,
            "arguments": args
        }, request_id=2)

        print(tool_call_response)

        # Parse and display the result
        result_json = parse_sse_response(tool_call_response)
        if result_json:
            print("\nParsed result:")
            print(json.dumps(result_json, indent=2))

# Try resources
print("\n\n=== Fetching Available Resources ===")
resources_response, _ = make_session_request("resources/list", request_id=3)
print(resources_response)

resources_json = parse_sse_response(resources_response)
if resources_json and "result" in resources_json:
    resources = resources_json.get("result", {}).get("resources", [])
    print(f"\nFound {len(resources)} resources:")
    for resource in resources:
        print(f"\n  URI: {resource.get('uri', 'N/A')}")
        print(f"  Name: {resource.get('name', 'N/A')}")
        print(f"  Description: {resource.get('description', 'N/A')}")
        print(f"  MIME Type: {resource.get('mimeType', 'N/A')}")

File: figma_mcp_client.js
#!/usr/bin/env node

const http = require('http');
const { Agent } = require('http');

const MCP_URL = 'http://127.0.0.1:3845/mcp';

let requestId = 0;
const agent = new Agent({ keepAlive: true });

function makeRequest(method, params = null) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify({
      jsonrpc: '2.0',
      id: ++requestId,
      method: method,
      ...(params && { params })
    });

    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': data.length,
        'Accept': 'application/json, text/event-stream',
        'Connection': 'keep-alive'
      },
      agent: agent
    };

    const req = http.request(MCP_URL, options, (res) => {
      let body = '';

      res.on('data', (chunk) => {
        body += chunk;
      });

      res.on('end', () => {
        // Parse SSE format
        if (body.includes('data: ')) {
          const lines = body.split('\n');
          for (const line of lines) {
            if (line.startsWith('data: ')) {
              try {
                const jsonData = JSON.parse(line.substring(6));
                resolve(jsonData);
                return;
              } catch (e) {
                // Continue to next line
              }
            }
          }
        }

        // Try parsing as regular JSON
        try {
          resolve(JSON.parse(body));
        } catch (e) {
          resolve({ raw: body });
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.write(data);
    req.end();
  });
}

// Helper to wait between requests
function wait(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function main() {
  try {
    // Initialize connection
    console.log('=== Initializing MCP Connection ===');
    const initResult = await makeRequest('initialize', {
      protocolVersion: '2024-11-05',
      capabilities: {},
      clientInfo: {
        name: 'claude-code-client',
        version: '1.0.0'
      }
    });
    console.log(JSON.stringify(initResult, null, 2));
    console.log();

    // Send initialized notification (no response expected)
    console.log('=== Sending initialized notification ===');
    const initializedNotification = await makeRequest('notifications/initialized');
    console.log(JSON.stringify(initializedNotification, null, 2));
    console.log();

    // Wait a bit for initialization
    await wait(100);

    // List tools
    console.log('=== Listing Available Tools ===');
    const toolsResult = await makeRequest('tools/list');
    console.log(JSON.stringify(toolsResult, null, 2));
    console.log();

    if (toolsResult.result && toolsResult.result.tools) {
      const tools = toolsResult.result.tools;
      console.log(`Found ${tools.length} tools:`);
      tools.forEach(tool => {
        console.log(`\n  - ${tool.name}`);
        console.log(`    Description: ${tool.description || 'No description'}`);
        if (tool.inputSchema) {
          console.log(`    Input Schema:`);
          console.log(JSON.stringify(tool.inputSchema, null, 6));
        }
      });

      // Call each tool to get design specs
      console.log('\n\n=== Calling Tools to Get Design Specifications ===');
      for (const tool of tools) {
        console.log(`\n--- Calling: ${tool.name} ---`);
        try {
          const result = await makeRequest('tools/call', {
            name: tool.name,
            arguments: {}
          });
          console.log(JSON.stringify(result, null, 2));
        } catch (error) {
          console.log(`Error calling ${tool.name}: ${error.message}`);
        }
      }
    }

    // List resources
    console.log('\n\n=== Listing Available Resources ===');
    const resourcesResult = await makeRequest('resources/list');
    console.log(JSON.stringify(resourcesResult, null, 2));

    if (resourcesResult.result && resourcesResult.result.resources) {
      const resources = resourcesResult.result.resources;
      console.log(`\nFound ${resources.length} resources:`);
      resources.forEach(resource => {
        console.log(`\n  - URI: ${resource.uri}`);
        console.log(`    Name: ${resource.name || 'N/A'}`);
        console.log(`    Description: ${resource.description || 'N/A'}`);
        console.log(`    MIME Type: ${resource.mimeType || 'N/A'}`);
      });

      // Try to read the first resource if available
      if (resources.length > 0) {
        console.log('\n\n=== Reading First Resource ===');
        try {
          const resourceData = await makeRequest('resources/read', {
            uri: resources[0].uri
          });
          console.log(JSON.stringify(resourceData, null, 2));
        } catch (error) {
          console.log(`Error reading resource: ${error.message}`);
        }
      }
    }

  } catch (error) {
    console.error('Error:', error);
  }
}

main();
